presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/MetadataQueryOptimizer.java

```diff
@@ -151,31 +166,111 @@ public PlanNode visitAggregation(AggregationNode node, RewriteContext<Void> cont
                 return context.defaultRewrite(node);
             }
 
-            ImmutableList.Builder<List<RowExpression>> rowsBuilder = ImmutableList.builder();
-            for (TupleDomain<ColumnHandle> domain : predicates.getPredicates()) {
-                if (!domain.isNone()) {
-                    Map<ColumnHandle, NullableValue> entries = TupleDomain.extractFixedValues(domain).get();
+            if (!isReducible(node)) {
```

Comment:
Maybe add a helper function when `isReducible(node)` is true and put the new logic there? This function is already pretty big.

Reply:
I extracted the newly added code into `reduce` (after changing `reduce` to `evaluateMinMax`). Also flipped the if statement so it looks like early return pattern.